metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java

```diff
@@ -445,28 +444,24 @@ public void testBalancePartitionLeaders() throws Throwable {
             active.alterPartition(ANONYMOUS_CONTEXT, new AlterPartitionRequest
                 .Builder(alterPartitionRequest, false).build((short) 0).data()).get();
 
-            AtomicLong lastHeartbeatMs = new AtomicLong(getMonotonicMs(active.time()));
+            AtomicLong lastHeartbeatNs = new AtomicLong(active.time().nanoseconds());
             sendBrokerHeartbeat(active, allBrokers, brokerEpochs);
             // Check that partitions are balanced
             TestUtils.waitForCondition(
                 () -> {
-                    long currentMonotonicMs = getMonotonicMs(active.time());
-                    if (currentMonotonicMs > lastHeartbeatMs.get() + (sessionTimeoutMillis / 2)) {
-                        lastHeartbeatMs.set(currentMonotonicMs);
```

Comment:
If `active.time().nanoseconds()` overflows (wraps), `currentMonotonicMs > lastHeartbeatMs.get() + (sessionTimeoutMillis / 2)` will be false until roughly 292 years (2^63 nanoseconds). The Java Doc recommends to not comparing `nanoTime()` and to instead compare elapsed time `nanoTime() - nanoTime()`. See the Java Doc for details: https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--